README.md
# Klient testing 

![badge-coverage](.github/badges/coverage.svg)

- [Introduction](#introduction)
- [Requirements](#requirements)
- [Installation](#installation)
- [MockAxiosWithRestApi](#mockaxioswithrestapi)

## Introduction

This repository purposes tools for testing [Klient library](https://github.com/klientjs/core) with [JEST](https://github.com/facebook/jest) to make test files more readable.


## Requirements

- [Node](https://github.com/nodejs/node) >= v12
- [Axios](https://github.com/axios/axios) >= v0.27.0


## Installation

Install testing package with your favorite package manager :

```shell
# With NPM
$ npm install --save-dev @klient/testing

# With YARN
$ yarn add --dev @klient/testing
```


## MockAxiosWithRestApi

Mock [Axios](https://github.com/axios/axios) with a fake Rest API described below.

```js
import axios from 'axios';
import { mockAxiosWithRestApi } from '@klient/testing';

//
// Mock Axios
//
jest.mock('axios');
const api = mockAxiosWithRestApi({
  // Optionally load Post resources data in API
  // posts: [
  //   {
  //     id: 'xxx',
  //     title: 'Title',
  //     content: 'Content',
  //     author: 'test'
  //   }
  // ]
});


//
// Start a test
//
test('example', async () => {
  let credentials = {};
  let postId = null;

  //
  // Get JWT token for protected routes
  //
  await axios
    .post('/auth', {
      username: 'test',
      password: 'test',
      // exp: '2h',         // Specify token duration
      // refreshExp: '3h'   // Specify refresh token duration
    })
    .then(response => {  
      console.log(response.data);
      // {
      //   "token": "...",
      //   "refresh_token": "..."
      // }

      credentials = response.data
    })
  ;


  //
  // Refresh JWT token with refresh token provided
  //
  await axios
    .post('/auth/refresh', { refresh_token: credentials.refresh_token })
    .then(response => {  
      console.log(response.data);
      // {
      //   "token": "...",
      //   "refresh_token": "..."
      // }

      credentials = response.data
    })
  ;


  //
  // Create a Post resource
  //
  // YOU MUST BE AUTHENTICATED WITH JWT TOKEN
  //
  await axios
    .post('/posts', { title: 'test', content: 'test' }, {
      headers: {
        Authorization: `Bearer ${credentials.token}`
      }
    })
    .then(response => {  
      console.log(response.data);
      // {
      //   "id": "2190428f-6785-41be-a4c6-e6ad7cd978de",
      //   "title": "test",
      //   "content": "test",
      //   "author": "test"
      // }

      postId = response.data.id;
    })
  ;


  //
  // Get Post resource collection
  //
  await axios
    .get('/posts')
    .then(response => {  
      console.log(response.data);
      // [
      //   {
      //     "id": "2190428f-6785-41be-a4c6-e6ad7cd978de",
      //     "title": "test",
      //     "content": "test",
      //     "author": "test"
      //   }
      // ]
    })
  ;


  //
  // Get Post resource item
  //
  await axios
    .get(`/posts/${postId}`)
    .then(response => {  
      console.log(response.data);
      // {
      //   "id": "2190428f-6785-41be-a4c6-e6ad7cd978de",
      //   "title": "test",
      //   "content": "test",
      //   "author": "test"
      // }
    })
  ;


  //
  // Update Post resource item
  //
  // YOU MUST BE AUTHENTICATED WITH JWT TOKEN FETCHED WITH THE SAME USERNAME AS POST AUTHOR
  //
  await axios
    .put(`/posts/${postId}`, { title: 'Test', content: 'Test' } {
      headers: {
        Authorization: `Bearer ${credentials.token}`
      }
    })
    .then(response => {  
      console.log(response.data);
      // {
      //   "id": "2190428f-6785-41be-a4c6-e6ad7cd978de",
      //   "title": "Test",
      //   "content": "Test",
      //   "author": "test"
      // }
    })
  ;


  //
  // Delete Post resource item
  //
  // YOU MUST BE AUTHENTICATED WITH JWT TOKEN FETCHED WITH THE SAME USERNAME AS POST AUTHOR
  //
  await axios
    .delete(`/posts/${postId}`, {
      headers: {
        Authorization: `Bearer ${credentials.token}`
      }
    })
  ;


  //
  // Not found
  //
  await axios
    .get('/unkown')
    .catch(e => {
      // Print 404
      console.log(e.response.status);
    })
  ;


  //
  // Reset API data
  //
  api.reset();
  await axios
    .get('/posts')
    .then(response => {
      console.log(e.response.data); // Print []
    })
  ;
});
```

**Notes**

  - The API will reject requests containing invalid token with 401 status code
  - The entrypoints whoses expect request body will return a 400 status code if invalid data are provided


CHANGELOG.md
# Changelog

# 1.1.0 (2023-05-14)


### Features

* mock axios with an api rest ([a810aec](https://github.com/klientjs/testing/commit/a810aec2af9835b630c98d03ea49e96292ce119c))

tests/api.test.ts
/**
 * @jest-environment jsdom
 */

import axios from 'axios';
import jwtDecode from 'jwt-decode';

import { mockAxiosWithRestApi } from '../src';
import RestApi from '../src/api/api';

let createdPostId: null | string = null;

let credentials = {
  token: '',
  refresh_token: ''
};

jest.mock('axios');

const username = 'user@localhost';
const api = mockAxiosWithRestApi({
  posts: [
    {
      id: 'xxx',
      title: 'Title',
      content: 'Content',
      author: username
    }
  ]
});

test('constructor', () => {
  expect(new RestApi()).toBeInstanceOf(RestApi);
});

test('login', () =>
  axios({
    url: '/auth',
    method: 'POST',
    data: {
      username,
      password: 'test',
      exp: '2h',
      refreshExp: '3h'
    }
  })
    .then(({ data, status }) => {
      expect(status).toBe(200);
      expect(data.token).toBeDefined();
      expect(data.refresh_token).toBeDefined();

      const tokenData: { username: string; exp: number } = jwtDecode(data.token);
      const refreshTokenData: { username: string; exp: number } = jwtDecode(data.refresh_token);

      expect(tokenData.username).toBe(username);
      expect(tokenData.exp).toBeGreaterThanOrEqual(Math.floor(Date.now() / 1000) + 7198);
      expect(tokenData.exp).toBeLessThanOrEqual(Math.floor(Date.now() / 1000) + 7200);

      expect(refreshTokenData.username).toBe(username);
      expect(refreshTokenData.exp).toBeGreaterThanOrEqual(Math.floor(Date.now() / 1000) + 10798);
      expect(refreshTokenData.exp).toBeLessThanOrEqual(Math.floor(Date.now() / 1000) + 10800);

      credentials = data;
    })
    .catch((e) => {
      console.log(e);
      throw e;
    }));

test('login:failure', () =>
  axios
    .post('/auth', { username })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(400);
    }));

test('login:refresh', () =>
  axios
    .request({
      url: '/auth/refresh',
      method: 'POST',
      data: {
        refresh_token: credentials.refresh_token
      }
    })
    .then(({ data, status }) => {
      expect(status).toBe(200);
      expect(data.token).toBeDefined();
      expect(data.refresh_token).toBeDefined();

      credentials = data;
    })
    .catch((e) => {
      console.log(e);
      throw e;
    }));

test('login:refresh:expired', async () => {
  let customCredentials = {
    token: '',
    refresh_token: ''
  };

  await axios
    .request({
      url: '/auth',
      method: 'POST',
      data: {
        username,
        password: 'test',
        exp: '0s',
        refreshExp: '0s'
      }
    })
    .then(({ data, status }) => {
      expect(status).toBe(200);
      expect(data.token).toBeDefined();
      expect(data.refresh_token).toBeDefined();

      customCredentials = data;
    })
    .catch((e) => {
      console.log(e);
      throw e;
    });

  await axios
    .request({
      url: '/auth/refresh',
      method: 'POST',
      data: {
        refresh_token: customCredentials.refresh_token
      }
    })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(400);
    });
});

test('login:refresh:failure', () =>
  axios
    .request({
      url: '/auth/refresh',
      method: 'POST',
      data: {
        refresh_token: 'unknown'
      }
    })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(400);
    }));

test('auth', () =>
  axios
    .request({
      url: '/posts',
      method: 'POST',
      data: {
        title: 'test',
        content: 'test'
      }
    })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(401);
      expect(e.response.data.error).toBe('Access denied.');
    }));

test('auth:invalid', async () => {
  await axios
    .get('/posts', {
      headers: {
        Authorization: 'Bearer FAKE'
      }
    })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(401);
    });
});

test('auth:expired', async () => {
  let customCredentials = {
    token: '',
    refresh_token: ''
  };

  await axios
    .request({
      url: '/auth',
      method: 'POST',
      data: {
        username,
        password: 'test',
        exp: '0s',
        refreshExp: '0s'
      }
    })
    .then(({ data, status }) => {
      expect(status).toBe(200);
      expect(data.token).toBeDefined();
      expect(data.refresh_token).toBeDefined();

      customCredentials = data;
    })
    .catch((e) => {
      console.log(e);
      throw e;
    });

  await axios
    .request({
      url: '/posts',
      method: 'GET',
      headers: {
        Authorization: `Bearer ${customCredentials.token}`
      }
    })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(401);
    });
});

test('resource:create', () => {
  const requests: Promise<unknown>[] = [];

  requests.push(
    axios
      .request({
        url: '/posts',
        method: 'POST',
        headers: {
          Authorization: `Bearer ${credentials.token}`
        },
        data: {
          title: 'Title created',
          content: 'Content created'
        }
      })
      .then(({ data, status }) => {
        expect(status).toBe(201);
        expect(data.id).toBeDefined();
        expect(data.title).toBe('Title created');
        expect(data.content).toBe('Content created');
        expect(data.author).toBe(username);

        createdPostId = data.id;
      })
      .catch((e) => {
        console.log(e);
        throw e;
      })
  );

  return Promise.all(requests);
});

test('resource:create:invalid', () =>
  axios
    .request({
      url: '/posts',
      method: 'POST',
      headers: {
        Authorization: `Bearer ${credentials.token}`
      },
      data: {
        title: 'test'
        // content: "test",
      }
    })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(400);
    }));

test('resource:read', () =>
  axios
    .request({
      url: `/posts/${createdPostId}`,
      method: 'GET'
    })
    .then(({ data, status }) => {
      expect(status).toBe(200);
      expect(data.id).toBe(createdPostId);
      expect(data.title).toBe('Title created');
      expect(data.content).toBe('Content created');
      expect(data.author).toBe(username);
    })
    .catch((e) => {
      console.log(e);
      throw e;
    }));

test('resource:read:collection', () =>
  axios
    .request({
      url: '/posts',
      method: 'GET'
    })
    .then(({ data, status }) => {
      expect(status).toBe(200);
      expect(data).toBeInstanceOf(Array);
      expect(data.length).toBe(2);

      expect(data[0].id).toBe('xxx');
      expect(data[0].title).toBe('Title');
      expect(data[0].content).toBe('Content');
      expect(data[0].author).toBe(username);

      expect(data[1].id).toBe(createdPostId);
      expect(data[1].title).toBe('Title created');
      expect(data[1].content).toBe('Content created');
      expect(data[1].author).toBe(username);
    })
    .catch((e) => {
      console.log(e);
      throw e;
    }));

test('resource:read:collection:filtered', () =>
  axios
    .request({
      url: '/posts',
      method: 'GET',
      params: {
        title: 'Title created',
        content: 'Content created'
      }
    })
    .then(({ data, status }) => {
      expect(status).toBe(200);
      expect(data).toBeInstanceOf(Array);
      expect(data.length).toBe(1);
      expect(data[0].title).toBe('Title created');
      expect(data[0].content).toBe('Content created');
    })
    .catch((e) => {
      console.log(e);
      throw e;
    }));

test('resource:update', () =>
  axios
    .put(
      `/posts/${createdPostId}`,
      {
        title: 'Title updated',
        content: 'Content updated'
      },
      {
        headers: {
          Authorization: `Bearer ${credentials.token}`
        }
      }
    )
    .then(({ data, status }) => {
      expect(status).toBe(200);
      expect(data.id).toBe(createdPostId);
      expect(data.title).toBe('Title updated');
      expect(data.content).toBe('Content updated');
      expect(data.author).toBe(username);
    })
    .catch((e) => {
      console.log(e);
      throw e;
    }));

test('resource:update:forbidden', async () => {
  let customCredentials = {
    token: ''
  };

  await axios
    .request({
      url: '/auth',
      method: 'POST',
      data: {
        username: 'unknow@localhost',
        password: 'unknow'
      }
    })
    .then(({ data }) => {
      customCredentials = data;
    });

  console.log(createdPostId);
  await axios
    .request({
      url: `/posts/${createdPostId}`,
      method: 'PUT',
      headers: {
        Authorization: `Bearer ${customCredentials.token}`
      },
      data: {
        title: '...',
        content: '...'
      }
    })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(403);
    });
});

test('resource:delete:forbidden', async () => {
  let customCredentials = {
    token: ''
  };

  await axios
    .request({
      url: '/auth',
      method: 'POST',
      data: {
        username: 'unknow@localhost',
        password: 'unknow'
      }
    })
    .then(({ data }) => {
      customCredentials = data;
    });

  await axios
    .request({
      url: `/posts/${createdPostId}`,
      method: 'DELETE',
      headers: {
        Authorization: `Bearer ${customCredentials.token}`
      }
    })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(403);
    });
});

test('resource:delete', () =>
  axios
    .request({
      url: `/posts/${createdPostId}`,
      method: 'DELETE',
      headers: {
        Authorization: `Bearer ${credentials.token}`
      }
    })
    .then(({ data, status }) => {
      expect(status).toBe(204);
      expect(data).toBeNull();
    })
    .catch((e) => {
      console.log(e);
      throw e;
    }));

test('notfound', () => {
  axios
    .put(
      '/posts/unknown',
      {},
      {
        headers: {
          Authorization: `Bearer ${credentials.token}`
        }
      }
    )
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(404);
    });

  axios
    .patch(
      '/posts/1',
      {},
      {
        headers: {
          Authorization: `Bearer ${credentials.token}`
        }
      }
    )
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(404);
    });

  axios
    .delete('/posts/unknown', {
      headers: {
        Authorization: `Bearer ${credentials.token}`
      }
    })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(404);
    });

  axios
    .get('/posts/unknown')
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(404);
    });

  axios({})
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(404);
    });
});

test('reset', () => {
  api.reset();

  axios
    .request({
      url: '/posts'
    })
    .then(({ data }) => {
      expect(data).toBeInstanceOf(Array);
      expect(data.length).toBe(0);
    })
    .catch((e) => {
      console.log(e);
      throw e;
    });
});

test('cancel', () => {
  const controller = new AbortController();

  controller.abort();

  return axios
    .request({
      url: '/posts',
      signal: controller.signal
    })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(axios.isCancel(e)).toBe(true);
    });
});

test('file', () => {
  return axios
    .request({
      url: '/file',
      responseType: 'blob'
    })
    .then(({ headers, data, status }) => {
      expect(status).toBe(200);
      expect(data).toBeInstanceOf(Blob);
      expect(headers['Content-Type']).toBe('text/plain');
      expect(headers['Content-Disposition']).toBeDefined();
    })
    .catch((e) => {
      console.log(e);
      throw e;
    });
});


src/api/helper.ts
const HttpStatusCode: { [_prop: string | number]: string | number } = {
  Continue: 100,
  SwitchingProtocols: 101,
  Processing: 102,
  EarlyHints: 103,
  Ok: 200,
  Created: 201,
  Accepted: 202,
  NonAuthoritativeInformation: 203,
  NoContent: 204,
  ResetContent: 205,
  PartialContent: 206,
  MultiStatus: 207,
  AlreadyReported: 208,
  ImUsed: 226,
  MultipleChoices: 300,
  MovedPermanently: 301,
  Found: 302,
  SeeOther: 303,
  NotModified: 304,
  UseProxy: 305,
  Unused: 306,
  TemporaryRedirect: 307,
  PermanentRedirect: 308,
  BadRequest: 400,
  Unauthorized: 401,
  PaymentRequired: 402,
  Forbidden: 403,
  NotFound: 404,
  MethodNotAllowed: 405,
  NotAcceptable: 406,
  ProxyAuthenticationRequired: 407,
  RequestTimeout: 408,
  Conflict: 409,
  Gone: 410,
  LengthRequired: 411,
  PreconditionFailed: 412,
  PayloadTooLarge: 413,
  UriTooLong: 414,
  UnsupportedMediaType: 415,
  RangeNotSatisfiable: 416,
  ExpectationFailed: 417,
  ImATeapot: 418,
  MisdirectedRequest: 421,
  UnprocessableEntity: 422,
  Locked: 423,
  FailedDependency: 424,
  TooEarly: 425,
  UpgradeRequired: 426,
  PreconditionRequired: 428,
  TooManyRequests: 429,
  RequestHeaderFieldsTooLarge: 431,
  UnavailableForLegalReasons: 451,
  InternalServerError: 500,
  NotImplemented: 501,
  BadGateway: 502,
  ServiceUnavailable: 503,
  GatewayTimeout: 504,
  HttpVersionNotSupported: 505,
  VariantAlsoNegotiates: 506,
  InsufficientStorage: 507,
  LoopDetected: 508,
  NotExtended: 510,
  NetworkAuthenticationRequired: 511
};

Object.entries(HttpStatusCode).forEach(([key, value]) => {
  HttpStatusCode[value] = key;
});

export default {
  getStatusText(status: number) {
    return String(HttpStatusCode[status]);
  }
};


src/api/router.ts
import type { AxiosRequestConfig } from 'axios';

type AnyObject = { [_prop: string]: any }; // eslint-disable-line

export interface Route {
  url: string;
  controller: string;
  method?: string;
  protected?: boolean;
}

export default class Router {
  constructor(protected routes: Route[]) {}

  match(request: AxiosRequestConfig): [Route, unknown] {
    const requestUrl = request.url || '';
    const requestMethod = request.method || 'GET';

    let target: Route = { url: '*', controller: 'notFound' };
    let parameters: AnyObject = {};

    for (let i = 0, len = this.routes.length; i < len; i += 1) {
      const route = this.routes[i];
      const routeUrl = route.url;
      const routeMethod = route.method || 'GET';

      // Method doesn't match
      if (routeMethod !== requestMethod) {
        continue;
      }

      const routePaths = routeUrl.split('/').filter((v) => typeof v === 'string' && v !== '');
      const requestPaths = requestUrl.split('/').filter((v) => typeof v === 'string' && v !== '');

      // Url parts doesn't match
      if (routePaths.length !== requestPaths.length) {
        continue;
      }

      let match = false;

      for (let n = 0, rlen = routePaths.length; n < rlen; n += 1) {
        const routePath = routePaths[n];
        const requestPath = requestPaths[n];

        if (routePath === requestPath) {
          match = n === rlen - 1;
          continue;
        }

        if (routePath.indexOf('{') === 0 && requestPath !== undefined) {
          const paramName = String(routePath).replace('{', '').replace('}', '');
          parameters[paramName] = requestPath;
          match = n === rlen - 1;
        }
      }

      if (match) {
        target = route;
        break;
      } else {
        parameters = {};
      }
    }

    return [target, parameters];
  }
}


src/api/api.ts
/* eslint-disable class-methods-use-this */

import { AxiosError, CanceledError } from 'axios';
import { v4 as uuidV4 } from 'uuid';
import jwtDecode from 'jwt-decode';
import * as jwt from 'jsonwebtoken';

import type { AxiosResponse, AxiosRequestConfig } from 'axios';
import helper from './helper';
import Router from './router';
import routes from './routes';

import type { Route } from './router';

export interface Post {
  id: string | null | undefined;
  title: string;
  content: string;
  author: string | null;
}

type AnyObject = any; // eslint-disable-line
type RouteParameters = AnyObject;
type ControllerResponse = {
  status: number;
  data: AnyObject;
  headers?: AnyObject;
};

export type RestApiProps = {
  posts?: Post[];
};

export default class RestApi {
  protected router = new Router(routes);

  protected posts: { [_id: string]: Post } = {};

  protected user: string | null = null;

  protected request: AxiosRequestConfig = {};

  protected tokens: string[] = [];

  protected refreshTokens: string[] = [];

  constructor(props?: RestApiProps) {
    if (props?.posts) {
      props.posts.forEach((post) => {
        this.posts[String(post.id)] = post;
      });
    }
  }

  reset() {
    this.clean();
    this.tokens = [];
    this.refreshTokens = [];
    this.posts = {};
  }

  clean() {
    this.user = null;
    this.request = {};
  }

  handle(request: AxiosRequestConfig): Promise<AxiosResponse> {
    if (request.signal?.aborted) {
      return this.sendCancelledResponse();
    }

    const [route, parameters] = this.router.match(request);

    // Check request and auth user
    const failureResponse = this.securizeRequest(request, route);

    // Invalid request
    if (failureResponse) {
      return this.send(failureResponse);
    }

    // Update current request
    this.request = request;

    // Call method to get response
    const response = (this as AnyObject)[route.controller](parameters);

    // Send formatted response
    return this.send(response);
  }

  protected auth() {
    const { username, password, exp, refreshExp } = this.request.data;

    if (!username || !password) {
      return this.badRequest('Invalid credentials.');
    }

    return this.ok({
      token: this.createToken(username, exp || '1d'),
      refresh_token: this.createToken(username, refreshExp || '2d', true)
    });
  }

  protected authRefresh() {
    const refreshToken = this.request.data.refresh_token;

    if (!refreshToken || !this.refreshTokens.includes(refreshToken)) {
      return this.badRequest('Invalid refresh token provided.');
    }

    if (this.isTokenExpired(refreshToken)) {
      return this.badRequest('Expired refresh token provided.');
    }

    const { username }: { username: string } = jwtDecode(refreshToken);

    this.refreshTokens.splice(this.refreshTokens.indexOf(refreshToken), 1);

    return this.ok({
      token: this.createToken(username, '1d'),
      refresh_token: this.createToken(username, '2d', true)
    });
  }

  protected readPost(parameters: RouteParameters) {
    const post = this.posts[String(parameters.id)];

    if (!post) {
      return this.notFound();
    }

    return this.ok(post);
  }

  protected createPost() {
    const { data } = this.request;
    const required = ['title', 'content'];

    for (let i = 0, len = required.length; i < len; i += 1) {
      if (!data || typeof data[required[i]] !== 'string') {
        return this.badRequest(`A string value is required for "${required[i]}" property`);
      }
    }

    const post = {
      id: uuidV4(),
      title: data.title,
      content: data.content,
      author: this.user
    };

    this.posts[post.id] = post;

    return this.created(post);
  }

  protected readPosts() {
    const { title, content } = this.request.params || {};

    let collection: Post[] = Object.keys(this.posts).map((id) => this.posts[id]);

    if (title) {
      collection = collection.filter((post) => post.title.indexOf(title) !== -1);
    }

    if (content) {
      collection = collection.filter((post) => post.content.indexOf(content) !== -1);
    }

    return this.ok(collection);
  }

  protected updatePost(parameters: RouteParameters) {
    const post = this.posts[String(parameters.id)];

    if (!post) {
      return this.notFound();
    }

    if (post.author !== this.user) {
      return this.forbidden();
    }

    const { data } = this.request;

    if (data && typeof data.title === 'string') {
      post.title = data.title;
    }

    if (data && typeof data.content === 'string') {
      post.content = data.content;
    }

    return this.ok(post);
  }

  protected deletePost(parameters: RouteParameters) {
    const post = this.posts[String(parameters.id)];

    if (!post) {
      return this.notFound();
    }

    if (post.author !== this.user) {
      return this.forbidden();
    }

    delete this.posts[String(post.id)];

    return this.noContent();
  }

  protected file() {
    const content = 'Hello world !';
    let data: string | Blob = content;

    if (this.request.responseType === 'blob') {
      data = new Blob([content], { type: 'text/plain' });
    }

    return {
      headers: {
        'Content-Disposition': 'attachment;filename=hello.txt',
        'Content-Type': 'text/plain'
      },
      status: 200,
      data
    };
  }

  protected notFound() {
    return {
      status: 404,
      data: {
        error: 'Not Found.'
      }
    };
  }

  protected accessDenied(message?: string) {
    return {
      status: 401,
      data: {
        error: message || 'Access denied.'
      }
    };
  }

  protected forbidden() {
    return {
      status: 403,
      data: {
        error: 'Forbidden.'
      }
    };
  }

  protected badRequest(message: string) {
    return {
      status: 400,
      data: {
        error: message
      }
    };
  }

  protected ok(data: unknown) {
    return {
      status: 200,
      data
    };
  }

  protected created(data: unknown) {
    return {
      status: 201,
      data
    };
  }

  protected noContent() {
    return {
      status: 204,
      data: null
    };
  }

  protected send(res: ControllerResponse): Promise<AxiosResponse> {
    const response = {
      headers: {
        'Content-Type': 'application/json',
        ...res.headers
      },
      statusText: helper.getStatusText(res.status),
      config: this.request,
      ...res
    };

    this.clean();

    if (response.status >= 200 && response.status < 400) {
      return Promise.resolve(response);
    }

    const err = new AxiosError();

    err.message = response.statusText;
    err.code = AxiosError.ERR_BAD_RESPONSE;
    err.config = this.request;
    err.response = response;

    return Promise.reject(err);
  }

  protected sendCancelledResponse() {
    const err = new CanceledError();

    Object.defineProperty(err, '__CANCEL__', { value: true });
    err.code = AxiosError.ERR_CANCELED;
    err.config = this.request;

    return Promise.reject(err);
  }

  protected securizeRequest(request: AxiosRequestConfig, route: Route) {
    const raw = request.headers?.Authorization;
    const token = typeof raw === 'string' ? raw.replace('Bearer ', '') : null;

    if (!token && route.protected) {
      return this.accessDenied();
    }

    if (token) {
      if (!this.tokens.includes(token)) {
        return this.accessDenied('Invalid token provided.');
      }

      const data: { exp: number; username: string } = jwtDecode(token);

      if (data.exp <= Math.floor(Date.now() / 1000)) {
        return this.accessDenied('Expired token provided.');
      }

      this.user = data.username;
    }
  }

  protected createToken(username: string, exp: string, isRefreshToken = false) {
    const token = jwt.sign({ username }, 'secret', { expiresIn: exp });
    this[isRefreshToken ? 'refreshTokens' : 'tokens'].push(token);
    return token;
  }

  protected isTokenExpired(token: string) {
    const data: { exp: number; username: string } = jwtDecode(token);
    return data.exp <= Math.floor(Date.now() / 1000);
  }
}
